### PR TITLE
Fix order of primitive basis for get_magnetic_symmetry

### LINKED ISSUE #370: Inconsistency in `primitive_lattice` results between `get_magnetic_symmetry_dataset` & `get_magnetic_symmetry` in the Python interface
I've discovered an inconsistency in Spglib: the `primitive_lattice` results from `get_magnetic_symmetry_dataset` and `get_magnetic_symmetry` do not match. Specifically, the latter's output seems to be the transpose of the former's. This issue is consistent across multiple tests, not just a single instance.

Here's a [brief code example](https://github.com/spglib/spglib/blob/8fab0b0a3272256aa3190181375009db82dd9b38/test/functional/python/test_magnetic_dataset.py#L51-L84) demonstrating the problem:

```python
>>> import numpy as np
>>> from spglib import get_magnetic_symmetry_dataset, get_magnetic_symmetry
>>> lattice = np.array(
        [
            [1.06949000e01, 0.00000000e00, 0.00000000e00],
            [3.84998337e-16, 6.28750000e00, 0.00000000e00],
            [3.09590711e-16, 3.09590711e-16, 5.05600000e00],
        ]
    )
>>> positions = np.array(
        [
            [0.0, 0.0, 0.0],
            [0.5, 0.0, 0.5],
            [0.5, 0.5, 0.5],
            [0.0, 0.5, 0.0],
            [0.2794, 0.25, 0.9906],
            [0.2206, 0.75, 0.4906],
            [0.7206, 0.75, 0.0094],
            [0.7794, 0.25, 0.5094],
        ]
    )
>>> numbers = [0, 0, 0, 0, 0, 0, 0, 0]
>>> magmoms = np.array(
        [
            [3.0, 0.4, 0.0],
            [-3.0, -0.4, 0.0],
            [-3.0, 0.4, 0.0],
            [3.0, -0.4, 0.0],
            [4.5, 0.0, 0.0],
            [-4.5, 0.0, 0.0],
            [4.5, 0.0, 0.0],
            [-4.5, 0.0, 0.0],
        ]
    )
>>> dataset = get_magnetic_symmetry_dataset((lattice, positions, numbers, magmoms))
>>> symdict = get_magnetic_symmetry((lattice, positions, numbers, magmoms))
>>> dataset
{'uni_number': 544, 'msg_type': 3, 'hall_number': 292, 'tensor_rank': 1, 'n_operations': 8, 'rotations': array([[[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0, -1]],

       [[-1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0,  1]],

       [[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0, -1]],

       [[ 1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0, -1]],

       [[-1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0, -1]],

       [[ 1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0,  1]]], dtype=int32), 'translations': array([[-0.00000000e+00, -0.00000000e+00, -0.00000000e+00],
       [-0.00000000e+00, -0.00000000e+00, -0.00000000e+00],
       [ 5.00000000e-01,  0.00000000e+00,  5.00000000e-01],
       [ 5.00000000e-01,  0.00000000e+00,  5.00000000e-01],
       [ 5.00000000e-01,  5.00000000e-01,  5.00000000e-01],
       [ 5.00000000e-01,  5.00000000e-01,  5.00000000e-01],
       [ 6.16297582e-33,  5.00000000e-01,  0.00000000e+00],
       [ 6.16297582e-33,  5.00000000e-01,  0.00000000e+00]]), 'time_reversals': array([False, False, False, False,  True,  True,  True,  True]), 'n_atoms': 8, 'equivalent_atoms': array([0, 0, 0, 0, 4, 4, 4, 4], dtype=int32), 'transformation_matrix': array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]]), 'origin_shift': array([0., 0., 0.]), 'n_std_atoms': 8, 'std_lattice': array([[1.06949000e+01, 0.00000000e+00, 0.00000000e+00],
       [3.84998337e-16, 6.28750000e+00, 0.00000000e+00],
       [3.09590711e-16, 3.09590711e-16, 5.05600000e+00]]), 'std_types': array([0, 0, 0, 0, 0, 0, 0, 0], dtype=int32), 'std_positions': array([[ 1.54074396e-33,  0.00000000e+00,  0.00000000e+00],
       [ 5.00000000e-01,  0.00000000e+00,  5.00000000e-01],
       [ 5.00000000e-01,  5.00000000e-01,  5.00000000e-01],
       [-4.62223187e-33,  5.00000000e-01,  0.00000000e+00],
       [ 2.79400000e-01,  2.50000000e-01,  9.90600000e-01],
       [ 2.20600000e-01,  7.50000000e-01,  4.90600000e-01],
       [ 7.20600000e-01,  7.50000000e-01,  9.40000000e-03],
       [ 7.79400000e-01,  2.50000000e-01,  5.09400000e-01]]), 'std_tensors': array([[ 3. ,  0.4,  0. ],
       [-3. , -0.4,  0. ],
       [-3. ,  0.4,  0. ],
       [ 3. , -0.4,  0. ],
       [ 4.5,  0. ,  0. ],
       [-4.5,  0. ,  0. ],
       [ 4.5,  0. ,  0. ],
       [-4.5,  0. ,  0. ]]), 'std_rotation_matrix': array([[1.00000000e+00, 1.23259516e-32, 0.00000000e+00],
       [0.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [0.00000000e+00, 0.00000000e+00, 1.00000000e+00]]), 'primitive_lattice': array([[-3.09590711e-16, -3.09590711e-16, -5.05600000e+00],
       [-3.84998337e-16, -6.28750000e+00, -0.00000000e+00],
       [-1.06949000e+01, -0.00000000e+00, -0.00000000e+00]])}
>>> symdict
{'rotations': array([[[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0, -1]],

       [[-1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0,  1]],

       [[ 1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0, -1]],

       [[ 1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0, -1]],

       [[-1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0,  1]],

       [[-1,  0,  0],
        [ 0,  1,  0],
        [ 0,  0, -1]],

       [[ 1,  0,  0],
        [ 0, -1,  0],
        [ 0,  0,  1]]], dtype=int32), 'translations': array([[-0.00000000e+00, -0.00000000e+00, -0.00000000e+00],
       [-0.00000000e+00, -0.00000000e+00, -0.00000000e+00],
       [ 5.00000000e-01,  0.00000000e+00,  5.00000000e-01],
       [ 5.00000000e-01,  0.00000000e+00,  5.00000000e-01],
       [ 5.00000000e-01,  5.00000000e-01,  5.00000000e-01],
       [ 5.00000000e-01,  5.00000000e-01,  5.00000000e-01],
       [ 6.16297582e-33,  5.00000000e-01,  0.00000000e+00],
       [ 6.16297582e-33,  5.00000000e-01,  0.00000000e+00]]), 'time_reversals': array([False, False, False, False,  True,  True,  True,  True]), 'equivalent_atoms': array([0, 0, 0, 0, 4, 4, 4, 4], dtype=int32), 'primitive_lattice': array([[-3.09590711e-16, -3.84998337e-16, -1.06949000e+01],
       [-3.09590711e-16, -6.28750000e+00, -0.00000000e+00],
       [-5.05600000e+00, -0.00000000e+00, -0.00000000e+00]])}
>>> symdict['primitive_lattice'] == dataset['primitive_lattice']
array([[ True, False, False],
       [False,  True,  True],
       [False,  True,  True]])
>>> symdict['primitive_lattice'] == dataset['primitive_lattice'].T
array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])
>>> import spglib
>>> spglib.get_version()
(2, 1, 0)
```

The comparison shows that `symdict['primitive_lattice']` is indeed the transpose of `dataset['primitive_lattice']`. This inconsistency might affect users needing accurate lattice information. I hope this information helps in resolving the issue.

### PR DESCRIPTION
**Pull request by lan496**
Fixes: https://github.com/spglib/spglib/issues/370

### CODE REVIEW

**Comment by lan496** [comment]
@atztogo This PR fixes my mistake of not transposing in Python API. I've checked other lattices like `std_lattice` are correctly handled.

File: python/spglib/spglib.py [hunk 1]
```diff
@@ -315,7 +315,7 @@ def get_magnetic_symmetry(
             "translations": np.array(translations[:num_sym], dtype="double", order="C"),
             "time_reversals": time_reversals,
             "equivalent_atoms": equivalent_atoms,
-            "primitive_lattice": primitive_lattice,
+            "primitive_lattice": np.transpose(primitive_lattice),
```
Comment (atztogo) [inline python/spglib/spglib.py]: What about below?
```python
"primitive_lattice": np.array(primitive_lattice.T, dtype="double", order="C")
```

**Comment by LecrisUT** [comment]
I think this needs a bit more rethinking. Why are the lattices transposed when both C and python use the same row-major?

Also, isn't all of the api supposed to go through `spg.dataset`?

**Comment by atztogo** [comment]
> I think this needs a bit more rethinking. Why are the lattices transposed when both C and python use the same row-major?

They are transposed each other. https://spglib.readthedocs.io/en/latest/python-spglib.html#crystal-structure-cell

This is due to historical reasons by my decisions.
There was no python interface initially. I prefer to have basis vectors in a 3x3 C-array resembling $(\mathbf{a} \ \mathbf{b} \ \mathbf{c})$ [lattice](https://spglib.readthedocs.io/en/latest/variable.html#lattice), i.e., a, b, c are put in the array like column vectors although I knew those in the row vectors are popular in condensed matter scientific codes.

I started to develop phonopy in python and so I wanted to make the python interface of spglib. I thought it was a good idea to follow the convention of the ASE's Atoms class, because at that time, python was not popular in science, but ASE and GPAW did good jobs in implementation in python. Of course I could never expecte that spglib becomes so popular, so I decided it without deep consideration. 

> Also, isn't all of the api supposed to go through `spg.dataset`?

It is done in C for space group and magnetic space group. But the transpose has to be made for basis vecotrs in python. We made this mistake due to complication introduced to have compatibility to my old implementation.

If I understand correctly, `get_magnetic_symmetry` is a successor of my old implementation existed in v1, and  `get_magnetic_symmetry_dataset`  is really new at v2. My original implementation had no ability to identify magnetic space group types and was not crystallographically well defined in the treatment of spins. @lan496 not only fixed the confusion of treatment of spins in my original implementation but also developed the identification of the magnetic space group types. Without comparing a set of symmetry operations with the magnetic space group type, it is not easy to obtain a set of symmetry operations that forms a group, i.e., my original implementation lacked this ability. Then @lan496 updated `get_magnetic_symmetry` to use the magnetic dataset as the back end to fix this issue.

**Comment by singularitti** [comment]
I also find the convention confusing, particularly the Python interface's need for lattice transposition. What's worse is that Julia uses column-major order, in contrast to C and Python's row-major order. This difference added an extra layer of confusion during the development of the Julia interface. To address this, I [wrote documentation](https://singularitti.github.io/Spglib.jl/stable/man/troubleshooting/#Returned-cell-symmetry-is-wrong) for my users. Nevertheless, I'm glad it's all sorted out.

> @lan496 not only fixed the confusion of treatment of spins in my original implementation but also developed the identification of the magnetic space group types. Without comparing a set of symmetry operations with the magnetic space group type, it is not easy to obtain a set of symmetry operations that forms a group, i.e., my original implementation lacked this ability.

Wow, that's some awesome work @lan496!

**Comment by atztogo** [comment]
If we have many wrappers, it may be good to have consensus about the definition of the basis vector matrix, where we can consider the python wrapper is an exception (because it is too late) and the others may, for example, follow the memory order in C. By this, in Fortran, it is transposed. However, in fact, we can introduce another confusion for atomic points either (num_points, 3) or (3, num_points). Finally @singularitti is right, documentation should ease the issue and developers have to be careful about it, if people think it is a better solution than developing an spglib-althernative with your favorite array definitions from scratch.

**Comment by lan496** [comment]
You may (already) realize column-major basis vectors are so natural in implementing crystallographic algorithms, but at the same time, we want to use row-major ones in the row-major preferred languages like Python 😇
For the Python interface, this PR will eliminate the issues about the order of basis vectors although it will not be a holistic solution. 

@atztogo Can I merge this PR?

**Comment by LecrisUT** [comment]
For the Python/C interface, the column/row major is not a problem, and the choice of column vectors is also perfectly fine. The issues here are rather:
- Inconsistency between Python/C interface if one version does the transpose and another doesn't. Regardless of the data-structure underneath, the mathematical notation `lattice[row][column]` is the same for all languages. This notation we must make sure is consistent 
- Making sure the ownership of the objects is consistent. In most cases, the wrapper should copy the content and make a native array of it. Using the transpose, would guarantee the data is copied

**Comment by atztogo** [comment]
> @atztogo Can I merge this PR?

Have you seen my comment above? Transpose of numpy array doesn't change data in memory, i.e., the order becomes 'F'. I prefer `order='C'` if the overhead to make it so is small.

**Comment by atztogo** [comment]
```
In [1]: import numpy as np

In [2]: a = np.arange(9).reshape(3,3)

In [3]: a
Out[3]:
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [4]: a.flags
Out[4]:
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

In [5]: np.transpose(a).flags
Out[5]:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

In [6]: np.array(a.T).flags
Out[6]:
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

In [7]: np.array(a.T, order='C').flags
Out[7]:
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

**Comment by lan496** [comment]
@atztogo Do you add comments on the code? If so, I cannot see it now. Anyway, I copied the array.

**Review by atztogo — CHANGES_REQUESTED** [review]

File: python/spglib/spglib.py (hunk 1 above)
Comment (lan496) [reply python/spglib/spglib.py]: Thanks. Now I can see this.

**Comment by atztogo** [comment]
Sorry, I didn't realize that I had to push the button to submit request the change.

**Review by atztogo — APPROVED** [review]

**Comment by lan496** [comment]
Not a problem at all. I have also had the same trouble before.

**Pull request merged**